Handle group membership for deactivated users.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
- First commit is to not include deactivated users in the members list.
- Second commit is to not include deactivated users in the mention data.
- Third commit is to not allow adding or removing deactivated users from groups.
- Fourth commit is to remove unnecessary check.
<!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

sahil839 commented on zerver/lib/user_groups.py:
```diff
@@ -293,13 +293,13 @@ def get_direct_user_groups(user_profile: UserProfile) -> List[UserGroup]:
 
 def get_user_group_direct_members(user_group: UserGroup) -> QuerySet[UserProfile]:
-    return user_group.direct_members.all()
+    return user_group.direct_members.filter(is_active=True)
```
This function is currently only callled in `is_user_in_group` and we can be sure that the user passed to it would be active, so filtering is not required here, but did this change for consistency and to make sure we do not encounter any bugs if this function is used at any other place in future.

timabbott replied:
Yeah definitely the safe approach.

sahil839 commented on zerver/lib/user_groups.py:
```diff
@@ -293,13 +293,13 @@ def get_direct_user_groups(user_profile: UserProfile) -> List[UserGroup]:
+    return user_group.direct_members.filter(is_active=True)
 
 
 def get_direct_memberships_of_users(user_group: UserGroup, members: List[UserProfile]) -> List[int]:
```
Have not added `user_profile__is_active=True` condition here since we expect the `members` list to only contain active users.

timabbott replied:
That seems OK. If you call this with a deactivated user, it seems like you . Could have a comment, like:

> Returns the subset of the provided `members` list who are direct subscribers of the group. If a deactivated user is passed, it will be returned if the user has was a member of the group when deactivated.

sahil839 commented on zerver/lib/mention.py:
```diff
@@ -271,7 +271,12 @@ def init_user_group_data(self, realm_id: int, content: str) -> None:
+                    "direct_members",
+                    queryset=UserProfile.objects.filter(realm_id=realm_id, is_active=True),
+                )
+            ):
```
Did this instead of using `group.direct_members.filter(is_active=True)` as that would defeat the purpose of using `prefetch_related`. This makes sure that the number of queries remain the same as it was before.

timabbott replied:
OK, please record the reason we're doing this in a comment...

Hmm, also, maybe we can tweak this to only prefetch the `ID` field, not the whole object and use a more efficient query? It looks like a `group.direct_members.all().values_list("id", flat=True)` might help? Would want to print-debug the SQL used to be sure, but I think this could be a significant little performance tweak.

sahil839 replied:
We cannot use `values_list` or `values` in the `queryset` arg to `Prefetch` and if we do `group.direct_members.all().values_list("id", flat=True)` then it generates an extra query.

timabbott replied:
OK, probably a comment is the right answer here.

sahil839 commented on zerver/lib/users.py:
```diff
@@ -204,13 +204,14 @@ def bulk_get_cross_realm_bots() -> Dict[str, UserProfile]:
-        UserProfile.objects.filter(id__in=user_ids, realm=realm).select_related("realm")
-    )
+def user_ids_to_users(
+    user_ids: Sequence[int], realm: Realm, *, allow_deactivated: bool = True
```
Not sure of the default value here. Currently `allow_deactivated` is always passed as `False`, but added the flag just for future use and as it was mentioned as TODO.

sahil839 commented:
We already do not show deactivated users in the new groups UI.

Regarding adding deactivated users to groups - deactivated users are not shown in typeahead, but someone can manually enter the email in input and the pill will be added to input. In the group edit form, deactivated user is ignored, but in group creation form the user is added in the members table and then the group creation fails on clicking on "Create". Stream creation form also has same behavior so will fix these in a separate PR.

sahil839 commented:
Rebased the PR and also fixed test coverage.

timabbott commented on zerver/actions/users.py:
```diff
@@ -385,6 +385,9 @@ def do_deactivate_user(
                 user_profile.realm, event_deactivate_bot, bot_owner_user_ids(user_profile)
             )
 
+        for user_group in get_direct_user_groups(user_profile):
```
I see, and `send_events_for_user_deactivation` has an early `return`, so it's a bit annoying to do it in there? I think it's actually fine for these to go in either order, though -- seeing a user removed from 4 groups and then deactivated as the sequences is perfectly coherent.

There's also a theory that we don't need to send events for this -- is removal from the group is implicit in a user being deactivated -- but maybe that's a pain for clients to implement. Probably worth discussing in "api design" though before we commit to changing this.

timabbott commented on zerver/views/user_groups.py:
```diff
@@ -202,9 +202,6 @@ def notify_for_user_group_subscription_changes(
             continue
-        if not recipient_user.is_active:
-            # Don't send notification message to deactivated users.
-            continue
```
Probably leaving an `assert not recipient_user.is_active` here would be a good idea, with a comment..

timabbott commented on zerver/lib/user_groups.py:
```diff
@@ -387,8 +387,10 @@ def user_groups_in_realm_serialized(realm: Realm) -> list[UserGroupDict]:
-        "user_group_id", "user_profile_id"
+    membership = (
+        UserGroupMembership.objects.filter(user_group__realm=realm)
+        .exclude(user_profile__is_active=False, user_group__named_user_group__isnull=False)
```
What's the story with the `user_group__named_user_group__isnull=` part here?

I'm finding it a bit hard to wrap my head around whether the exclude should be two separate excludes or a single one.

sahil839 replied:
Yeah, I wanted to ask this. This is for including the deactivated members in setting groups, i.e. anonymous groups. Ideally we would want to exclude deactivated users from members list for those as well, but that would need some more changes as well like to send the events to update the settings as well, so just wanted to confirm.

sahil839 commented:
Added a commit at the end for not including deactivated group members in the anonymous groups to get an idea about how the code would be.

timabbott commented on zerver/lib/events.py:
```diff
@@ -1112,6 +1113,22 @@ def _draft_update_action(i: int) -> None:
+                                group[setting_name].direct_members = [
+                                    user_id
+                                    for user_id in group[setting_name].direct_members
+                                    if user_id != person_user_id
```
Shouldn't there be something similar for Stream hanging settings?

sahil839 replied:
We currently do not have stream settings that support anonymous groups, so skipped that for now. But can add those as well.

timabbott replied:
We will in the future, so it seems like we should just have the logic as one less thing to remember when we do those. Though maybe `test_events` will fail for us naturally.

sahil839 commented:
Have not done the frontend changes yet. Can do that if the backend approach seems good.

timabbott commented:
I gave this a thorough read and the backend approach seems fine. I'm not sure if this can merged as-is without the frontend changes; I think it should be safe? If it makes sense to merge this for API feature level marking reasons and put the frontend changes in a separate PR, that seems OK to me.

timabbott commented:
Speaking of which, this does need API documentation / changelog updates.

sahil839 commented on zerver/openapi/zulip.yaml:
```diff
@@ -20550,6 +20563,10 @@ paths:
                               type: array
                               description: |
                                 The integer user IDs of the user group members.
```
I am not sure whether we want to mention here that only active users are included. Like we do not mention this for stream subscribers and it is kind of understood that only active users should be included.

timabbott replied:
I'll think on this. Probably just going to do a bonus copyediting commit.

sahil839 commented on zerver/openapi/zulip.yaml:
```diff
@@ -3182,9 +3182,16 @@ paths:
+                                user is direct member of. When deactivating the user, event
+                                is only sent to users who cannot access the deactivated user.
+
+                                **Changes**: Starting with Zulip 10.0 (feature level 303), this
```
Would be good to get suggestions to phrase this to make it clear about it only happens for settings set to anonymous groups.

sahil839 commented:
Added the documentation changes, but that would still need some rephrasing. Also, moved the commit with other backend commits.

We can do the frontend change later, only the settings would not be live-updated. I tried doing the frontend changes, but that will need a couple of refactor commits, so would be better to have a separate PR.
